### ISSUE TITLE
Handling control statements of any depth in `return_linter`

### ISSUE THREAD

**Issue by MEO265** [r-lib/lintr#2356]
I think this double lint is not correct either you see the `if` after an `else` as a new expression, then only the second one should be marked, or you understand an `if` with a lot of `else` as one expression (that's how I would see it) and then only the first one would have to be marked.

```R
lint(
  trim_some("
  function(x, y) {
    if(x) {
      1
    } else if(y) {
      2
    }
  }
  "),
  return_linter(return_style = "explicit", allow_implicit_else = FALSE)
)
```
```

<text>:2:9: style: [return_linter] All functions must have an explicit return().
  if(x) {
        ^
<text>:4:10: style: [return_linter] All functions must have an explicit return().
  } else if(y) {
         ^~~~~~~
```

I'm not entirely sure if it's a bug of this PR or the first one. If it's one of the first, feel free to convert this message into an issue, because then I'll probably have a (simple) solution to fix it and properly handle control statements of any depth.

_Originally posted by @MEO265 in https://github.com/r-lib/lintr/issues/2321#issuecomment-1827240302_

**Comment by MEO265** [comment]
This bug should be fixed at the same time

```R
lint(
  trim_some("
  function(x, y) {
    if(x) {
      1
    } else if(y) {
      2
    }
  }
  "),
  return_linter(return_style = "implicit", allow_implicit_else = FALSE)
)

# No lint, but should lint
```

https://github.com/r-lib/lintr/pull/2321#issuecomment-1827223428
